File: app/src/main/java/de/devmil/paperlaunch/view/fragments/SettingsFragment.java
/*
 * Copyright 2015 Devmil Solutions
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package de.devmil.paperlaunch.view.fragments;

import android.content.Context;
import android.os.Bundle;
import android.preference.CheckBoxPreference;
import android.preference.ListPreference;
import android.preference.Preference;
import android.preference.PreferenceCategory;
import android.preference.PreferenceFragment;
import android.preference.PreferenceScreen;
import android.util.DisplayMetrics;

import java.util.ArrayList;
import java.util.List;

import de.devmil.paperlaunch.R;
import de.devmil.paperlaunch.config.LauncherGravity;
import de.devmil.paperlaunch.config.UserSettings;
import de.devmil.paperlaunch.service.LauncherOverlayService;
import de.devmil.paperlaunch.view.preferences.SeekBarPreference;

public class SettingsFragment extends PreferenceFragment {

    private UserSettings mUserSettings;
    private IActivationParametersChangedListener mActivationParametersChangedListener;

    public interface IActivationParametersChangedListener {
        void onActivationParametersChanged();
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        Context context = getActivity();

        mUserSettings = new UserSettings(context);

        PreferenceScreen screen = getPreferenceManager().createPreferenceScreen(context);
        setPreferenceScreen(screen);

        addActivationSettings(context, screen);
        addAppearanceSettings(context, screen);
    }

    public void setOnActivationParametersChangedListener(IActivationParametersChangedListener listener) {
        mActivationParametersChangedListener = listener;
    }

    private void fireActivationParametersChanged() {
        if(mActivationParametersChangedListener != null) {
            mActivationParametersChangedListener.onActivationParametersChanged();
        }
    }

    private void addActivationSettings(Context context, PreferenceScreen screen) {
        PreferenceCategory activationCategory = new PreferenceCategory(context);

        screen.addPreference(activationCategory);

        activationCategory.setPersistent(false);
        activationCategory.setTitle(R.string.fragment_settings_category_activation_title);
        activationCategory.setIcon(R.mipmap.ic_wifi_tethering_black_36dp);

        SeekBarPreference sensitivityPreference = new SeekBarPreference(context, 5, 40);
        activationCategory.addPreference(sensitivityPreference);

        sensitivityPreference.setValue(mUserSettings.getSensitivityDip());
        sensitivityPreference.setTitle(R.string.fragment_settings_activation_sensitivity_title);
        sensitivityPreference.setPersistent(false);
        sensitivityPreference.setOnPreferenceChangeListener(new Preference.OnPreferenceChangeListener() {
            @Override
            public boolean onPreferenceChange(Preference preference, Object newValue) {
                mUserSettings.load(getActivity());
                mUserSettings.setSensitivityDip((Integer) newValue);
                mUserSettings.save(getActivity());
                fireActivationParametersChanged();
                return true;
            }
        });

        DisplayMetrics metrics = new DisplayMetrics();
        getActivity().getWindowManager().getDefaultDisplay().getMetrics(metrics);

        final float heightDpi = metrics.heightPixels / metrics.density;

        SeekBarPreference offsetHeightPreference = new SeekBarPreference(context, 0, (int)heightDpi);
        activationCategory.addPreference(offsetHeightPreference);

        offsetHeightPreference.setValue((int) heightDpi - mUserSettings.getActivationOffsetHeightDip());
        offsetHeightPreference.setTitle(R.string.fragment_settings_activation_offset_height_title);
        offsetHeightPreference.setPersistent(false);
        offsetHeightPreference.setOnPreferenceChangeListener(new Preference.OnPreferenceChangeListener() {
            @Override
            public boolean onPreferenceChange(Preference preference, Object newValue) {
                mUserSettings.load(getActivity());
                mUserSettings.setActivationOffsetHeightDip((int) heightDpi - (Integer) newValue);
                mUserSettings.save(getActivity());
                fireActivationParametersChanged();
                return true;
            }
        });

        int offsetMin = -(int)(heightDpi / 2);
        int offsetMax = (int)(heightDpi / 2);

        SeekBarPreference offsetPositionPreference = new SeekBarPreference(context, offsetMin, offsetMax);
        activationCategory.addPreference(offsetPositionPreference);

        offsetPositionPreference.setValue(mUserSettings.getActivationOffsetPositionDip());
        offsetPositionPreference.setTitle(R.string.fragment_settings_activation_offset_position_title);
        offsetPositionPreference.setPersistent(false);
        offsetPositionPreference.setOnPreferenceChangeListener(new Preference.OnPreferenceChangeListener() {
            @Override
            public boolean onPreferenceChange(Preference preference, Object newValue) {
                mUserSettings.load(getActivity());
                mUserSettings.setActivationOffsetPositionDip((Integer) newValue);
                mUserSettings.save(getActivity());
                fireActivationParametersChanged();
                return true;
            }
        });

        CheckBoxPreference vibrateOnActivationPreference = new CheckBoxPreference(context);
        activationCategory.addPreference(vibrateOnActivationPreference);

        vibrateOnActivationPreference.setChecked(mUserSettings.isVibrateOnActivation());
        vibrateOnActivationPreference.setTitle(R.string.fragment_settings_activation_vibrate_on_activation_title);
        vibrateOnActivationPreference.setSummaryOn(R.string.fragment_settings_activation_vibrate_on_activation_summary_on);
        vibrateOnActivationPreference.setSummaryOff(R.string.fragment_settings_activation_vibrate_on_activation_summary_off);
        vibrateOnActivationPreference.setPersistent(false);
        vibrateOnActivationPreference.setOnPreferenceChangeListener(new Preference.OnPreferenceChangeListener() {
            @Override
            public boolean onPreferenceChange(Preference preference, Object newValue) {
                mUserSettings.load(getActivity());
                mUserSettings.setVibrateOnActivation((Boolean) newValue);
                mUserSettings.save(getActivity());
                LauncherOverlayService.notifyConfigChanged(getActivity());
                return true;
            }
        });
    }

    private void addAppearanceSettings(final Context context, PreferenceScreen screen) {
        PreferenceCategory apperanceCategory = new PreferenceCategory(context);
        screen.addPreference(apperanceCategory);

        apperanceCategory.setPersistent(false);
        apperanceCategory.setTitle(R.string.fragment_settings_category_appearance_title);

        CheckBoxPreference showBackgroundPreference = new CheckBoxPreference(context);
        apperanceCategory.addPreference(showBackgroundPreference);

        showBackgroundPreference.setPersistent(false);
        showBackgroundPreference.setTitle(R.string.fragment_settings_appearance_background_title);
        showBackgroundPreference.setSummaryOn(R.string.fragment_settings_appearance_background_summary_on);
        showBackgroundPreference.setSummaryOff(R.string.fragment_settings_appearance_background_summary_off);
        showBackgroundPreference.setChecked(mUserSettings.isShowBackground());
        showBackgroundPreference.setOnPreferenceChangeListener(new Preference.OnPreferenceChangeListener() {
            @Override
            public boolean onPreferenceChange(Preference preference, Object newValue) {
                mUserSettings.load(getActivity());
                mUserSettings.setShowBackground((Boolean) newValue);
                mUserSettings.save(getActivity());
                LauncherOverlayService.notifyConfigChanged(getActivity());
                return true;
            }
        });

        final ListPreference sidePreference = new ListPreference(context);
        apperanceCategory.addPreference(sidePreference);

        class SideEntry {
            SideEntry(String title, boolean value) {
                this.title = title;
                this.value = value;
            }
            String title;
            boolean value;
        }

        SideEntry[] sideEntryArray = new SideEntry[]
                {
                        new SideEntry(
                                context.getString(R.string.fragment_settings_appearance_side_optionleft),
                                false
                        ),
                        new SideEntry(
                                context.getString(R.string.fragment_settings_appearance_side_optionright),
                                true
                        )
                };

        List<CharSequence> sideEntryTitles = new ArrayList<>();
        List<CharSequence> sideEntryValues = new ArrayList<>();
        for(SideEntry se : sideEntryArray) {
            sideEntryTitles.add(se.title);
            sideEntryValues.add(Boolean.toString(se.value));
        }

        sidePreference.setPersistent(false);
        sidePreference.setTitle(R.string.fragment_settings_appearance_side_title);
        sidePreference.setEntries(sideEntryTitles.toArray(new CharSequence[sideEntryTitles.size()]));
        sidePreference.setEntryValues(sideEntryValues.toArray(new CharSequence[sideEntryValues.size()]));
        sidePreference.setValue(Boolean.toString(mUserSettings.isOnRightSide()));
        sidePreference.setSummary(context.getString(getSideSummary(mUserSettings.isOnRightSide())));
        sidePreference.setOnPreferenceChangeListener(new Preference.OnPreferenceChangeListener() {
            @Override
            public boolean onPreferenceChange(Preference preference, Object newValue) {
                boolean newBooleanValue = Boolean.parseBoolean((String) newValue);
                mUserSettings.load(getActivity());
                mUserSettings.setIsOnRightSide(newBooleanValue);
                mUserSettings.save(getActivity());
                sidePreference.setSummary(context.getString(getSideSummary(mUserSettings.isOnRightSide())));
                fireActivationParametersChanged();
                return true;
            }
        });

        final ListPreference gravityPreference = new ListPreference(context);
        apperanceCategory.addPreference(gravityPreference);

        class GravityEntry {
            GravityEntry(String title, LauncherGravity value) {
                this.title = title;
                this.value = value;
            }
            String title;
            LauncherGravity value;
        }

        GravityEntry[] gravityEntryArray = new GravityEntry[]
                {
                        new GravityEntry(
                                context.getString(R.string.fragment_settings_appearance_gravity_optiontop),
                                LauncherGravity.Top
                        ),
                        new GravityEntry(
                                context.getString(R.string.fragment_settings_appearance_gravity_optioncenter),
                                LauncherGravity.Center
                        ),
                        new GravityEntry(
                                context.getString(R.string.fragment_settings_appearance_gravity_optionbottom),
                                LauncherGravity.Bottom
                        )
                };

        List<CharSequence> gravityEntryTitles = new ArrayList<>();
        List<CharSequence> gravityEntryValues = new ArrayList<>();
        for(GravityEntry ge : gravityEntryArray) {
            gravityEntryTitles.add(ge.title);
            gravityEntryValues.add(Integer.toString(ge.value.getValue()));
        }

        gravityPreference.setPersistent(false);
        gravityPreference.setTitle(R.string.fragment_settings_appearance_gravity_title);
        gravityPreference.setEntries(gravityEntryTitles.toArray(new CharSequence[gravityEntryTitles.size()]));
        gravityPreference.setEntryValues(gravityEntryValues.toArray(new CharSequence[gravityEntryValues.size()]));
        gravityPreference.setValue(Integer.toString(mUserSettings.getLauncherGravity().getValue()));
        gravityPreference.setSummary(context.getString(getGravitySummary(mUserSettings.getLauncherGravity())));
        gravityPreference.setOnPreferenceChangeListener(new Preference.OnPreferenceChangeListener() {
            @Override
            public boolean onPreferenceChange(Preference preference, Object newValue) {
                int newIntValue = Integer.parseInt((String) newValue);
                LauncherGravity gravityValue = LauncherGravity.fromValue(newIntValue);
                mUserSettings.load(getActivity());
                mUserSettings.setLauncherGravity(gravityValue);
                mUserSettings.save(getActivity());
                gravityPreference.setSummary(context.getString(getGravitySummary(mUserSettings.getLauncherGravity())));
                fireActivationParametersChanged();
                return true;
            }
        });
    }

    private int getSideSummary(boolean isOnRightSide) {
        return isOnRightSide ?
                R.string.fragment_settings_appearance_side_optionright_summary
                : R.string.fragment_settings_appearance_side_optionleft_summary;
    }

    private int getGravitySummary(LauncherGravity gravity) {
        switch(gravity) {
            case Top:
                return R.string.fragment_settings_appearance_gravity_optiontop_summary;
            case Center:
                return R.string.fragment_settings_appearance_gravity_optioncenter_summary;
            case Bottom:
                return R.string.fragment_settings_appearance_gravity_optionbottom_summary;
        }
        return -1;
    }
}

File: app/src/main/java/de/devmil/paperlaunch/view/fragments/SettingsFragment.kt
/*
 * Copyright 2015 Devmil Solutions
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package de.devmil.paperlaunch.view.fragments

import android.content.Context
import android.os.Bundle
import android.preference.CheckBoxPreference
import android.preference.ListPreference
import android.preference.Preference
import android.preference.PreferenceCategory
import android.preference.PreferenceFragment
import android.preference.PreferenceScreen
import android.util.DisplayMetrics

import java.util.ArrayList

import de.devmil.paperlaunch.R
import de.devmil.paperlaunch.config.LauncherGravity
import de.devmil.paperlaunch.config.UserSettings
import de.devmil.paperlaunch.service.LauncherOverlayService
import de.devmil.paperlaunch.view.preferences.SeekBarPreference

class SettingsFragment : PreferenceFragment() {

    private var userSettings: UserSettings? = null
    private var activationParametersChangedListener: (() -> Unit)? = null

    @Deprecated("Deprecated in Java")
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        val context = activity

        userSettings = UserSettings(context)

        val screen = preferenceManager.createPreferenceScreen(context)
        preferenceScreen = screen

        addActivationSettings(context, screen)
        addAppearanceSettings(context, screen)
    }

    fun setOnActivationParametersChangedListener(listener: () -> Unit) {
        activationParametersChangedListener = listener
    }

    private fun fireActivationParametersChanged() {
        if (activationParametersChangedListener != null) {
            activationParametersChangedListener!!()
        }
    }

    private fun addActivationSettings(context: Context, screen: PreferenceScreen) {
        val activationCategory = PreferenceCategory(context)

        screen.addPreference(activationCategory)

        activationCategory.isPersistent = false
        activationCategory.setTitle(R.string.fragment_settings_category_activation_title)
        activationCategory.setIcon(R.mipmap.ic_wifi_tethering_black_36dp)

        val sensitivityPreference = SeekBarPreference(context, 5, 40)
        activationCategory.addPreference(sensitivityPreference)

        sensitivityPreference.setValue(userSettings!!.sensitivityDip)
        sensitivityPreference.setTitle(R.string.fragment_settings_activation_sensitivity_title)
        sensitivityPreference.isPersistent = false
        sensitivityPreference.onPreferenceChangeListener = Preference.OnPreferenceChangeListener { _, newValue ->
            userSettings!!.load(activity)
            userSettings!!.sensitivityDip = newValue as Int
            userSettings!!.save(activity)
            fireActivationParametersChanged()
            true
        }

        val metrics = DisplayMetrics()
        activity.windowManager.defaultDisplay.getMetrics(metrics)

        val heightDpi = metrics.heightPixels / metrics.density

        val offsetHeightPreference = SeekBarPreference(context, 0, heightDpi.toInt())
        activationCategory.addPreference(offsetHeightPreference)

        offsetHeightPreference.setValue(heightDpi.toInt() - userSettings!!.activationOffsetHeightDip)
        offsetHeightPreference.setTitle(R.string.fragment_settings_activation_offset_height_title)
        offsetHeightPreference.isPersistent = false
        offsetHeightPreference.onPreferenceChangeListener = Preference.OnPreferenceChangeListener { _, newValue ->
            userSettings!!.load(activity)
            userSettings!!.activationOffsetHeightDip = heightDpi.toInt() - newValue as Int
            userSettings!!.save(activity)
            fireActivationParametersChanged()
            true
        }

        val offsetMin = -(heightDpi / 2).toInt()
        val offsetMax = (heightDpi / 2).toInt()

        val offsetPositionPreference = SeekBarPreference(context, offsetMin, offsetMax)
        activationCategory.addPreference(offsetPositionPreference)

        offsetPositionPreference.setValue(userSettings!!.activationOffsetPositionDip)
        offsetPositionPreference.setTitle(R.string.fragment_settings_activation_offset_position_title)
        offsetPositionPreference.isPersistent = false
        offsetPositionPreference.onPreferenceChangeListener = Preference.OnPreferenceChangeListener { _, newValue ->
            userSettings!!.load(activity)
            userSettings!!.activationOffsetPositionDip = newValue as Int
            userSettings!!.save(activity)
            fireActivationParametersChanged()
            true
        }

        val vibrateOnActivationPreference = CheckBoxPreference(context)
        activationCategory.addPreference(vibrateOnActivationPreference)

        vibrateOnActivationPreference.isChecked = userSettings!!.isVibrateOnActivation
        vibrateOnActivationPreference.setTitle(R.string.fragment_settings_activation_vibrate_on_activation_title)
        vibrateOnActivationPreference.setSummaryOn(R.string.fragment_settings_activation_vibrate_on_activation_summary_on)
        vibrateOnActivationPreference.setSummaryOff(R.string.fragment_settings_activation_vibrate_on_activation_summary_off)
        vibrateOnActivationPreference.isPersistent = false
        vibrateOnActivationPreference.onPreferenceChangeListener = Preference.OnPreferenceChangeListener { _, newValue ->
            userSettings!!.load(activity)
            userSettings!!.isVibrateOnActivation = newValue as Boolean
            userSettings!!.save(activity)
            LauncherOverlayService.notifyConfigChanged(activity)
            true
        }
    }

    private fun addAppearanceSettings(context: Context, screen: PreferenceScreen) {
        val apperanceCategory = PreferenceCategory(context)
        screen.addPreference(apperanceCategory)

        apperanceCategory.isPersistent = false
        apperanceCategory.setTitle(R.string.fragment_settings_category_appearance_title)

        val showBackgroundPreference = CheckBoxPreference(context)
        apperanceCategory.addPreference(showBackgroundPreference)

        showBackgroundPreference.isPersistent = false
        showBackgroundPreference.setTitle(R.string.fragment_settings_appearance_background_title)
        showBackgroundPreference.setSummaryOn(R.string.fragment_settings_appearance_background_summary_on)
        showBackgroundPreference.setSummaryOff(R.string.fragment_settings_appearance_background_summary_off)
        showBackgroundPreference.isChecked = userSettings!!.isShowBackground
        showBackgroundPreference.onPreferenceChangeListener = Preference.OnPreferenceChangeListener { _, newValue ->
            userSettings!!.load(activity)
            userSettings!!.isShowBackground = newValue as Boolean
            userSettings!!.save(activity)
            LauncherOverlayService.notifyConfigChanged(activity)
            true
        }

        val sidePreference = ListPreference(context)
        apperanceCategory.addPreference(sidePreference)

        class SideEntry(var title: String, var value: Boolean)

        val sideEntryArray = arrayOf(SideEntry(
                context.getString(R.string.fragment_settings_appearance_side_optionleft),
                false
        ), SideEntry(
                context.getString(R.string.fragment_settings_appearance_side_optionright),
                true
        ))

        val sideEntryTitles = ArrayList<CharSequence>()
        val sideEntryValues = ArrayList<CharSequence>()
        for (se in sideEntryArray) {
            sideEntryTitles.add(se.title)
            sideEntryValues.add(java.lang.Boolean.toString(se.value))
        }

        sidePreference.isPersistent = false
        sidePreference.setTitle(R.string.fragment_settings_appearance_side_title)
        sidePreference.entries = sideEntryTitles.toTypedArray()
        sidePreference.entryValues = sideEntryValues.toTypedArray()
        sidePreference.value = java.lang.Boolean.toString(userSettings!!.isOnRightSide)
        sidePreference.summary = context.getString(getSideSummary(userSettings!!.isOnRightSide))
        sidePreference.onPreferenceChangeListener = Preference.OnPreferenceChangeListener { _, newValue ->
            val newBooleanValue = java.lang.Boolean.parseBoolean(newValue as String)
            userSettings!!.load(activity)
            userSettings!!.isOnRightSide = newBooleanValue
            userSettings!!.save(activity)
            sidePreference.summary = context.getString(getSideSummary(userSettings!!.isOnRightSide))
            fireActivationParametersChanged()
            true
        }

        val gravityPreference = ListPreference(context)
        apperanceCategory.addPreference(gravityPreference)

        class GravityEntry(var title: String, var value: LauncherGravity)

        val gravityEntryArray = arrayOf(GravityEntry(
                context.getString(R.string.fragment_settings_appearance_gravity_optiontop),
                LauncherGravity.Top
        ), GravityEntry(
                context.getString(R.string.fragment_settings_appearance_gravity_optioncenter),
                LauncherGravity.Center
        ), GravityEntry(
                context.getString(R.string.fragment_settings_appearance_gravity_optionbottom),
                LauncherGravity.Bottom
        ))

        val gravityEntryTitles = ArrayList<CharSequence>()
        val gravityEntryValues = ArrayList<CharSequence>()
        for (ge in gravityEntryArray) {
            gravityEntryTitles.add(ge.title)
            gravityEntryValues.add(Integer.toString(ge.value.value))
        }

        gravityPreference.isPersistent = false
        gravityPreference.setTitle(R.string.fragment_settings_appearance_gravity_title)
        gravityPreference.entries = gravityEntryTitles.toTypedArray()
        gravityPreference.entryValues = gravityEntryValues.toTypedArray()
        gravityPreference.value = Integer.toString(userSettings!!.launcherGravity.value)
        gravityPreference.summary = context.getString(getGravitySummary(userSettings!!.launcherGravity))
        gravityPreference.onPreferenceChangeListener = Preference.OnPreferenceChangeListener { _, newValue ->
            val newIntValue = Integer.parseInt(newValue as String)
            val gravityValue = LauncherGravity.fromValue(newIntValue)
            userSettings!!.load(activity)
            userSettings!!.launcherGravity = gravityValue
            userSettings!!.save(activity)
            gravityPreference.summary = context.getString(getGravitySummary(userSettings!!.launcherGravity))
            fireActivationParametersChanged()
            true
        }

        val itemScalePreference = SeekBarPreference(context, 50, 150)
        apperanceCategory.addPreference(itemScalePreference)

        itemScalePreference.setValue(userSettings!!.itemScalePercent)
        itemScalePreference.setTitle(R.string.fragment_settings_appearance_item_scale_title)
        itemScalePreference.isPersistent = false
        itemScalePreference.onPreferenceChangeListener = Preference.OnPreferenceChangeListener { _, newValue ->
            userSettings!!.load(activity)
            userSettings!!.itemScalePercent = newValue as Int
            userSettings!!.save(activity)
            LauncherOverlayService.notifyDataConfigChanged(activity)
            true
        }

        val showLogoPreference = CheckBoxPreference(context)
        apperanceCategory.addPreference(showLogoPreference)

        showLogoPreference.isChecked = userSettings!!.showLogo
        showLogoPreference.setTitle(R.string.fragment_settings_activation_show_logo_title)
        showLogoPreference.setSummaryOn(R.string.fragment_settings_activation_show_logo_summary_on)
        showLogoPreference.setSummaryOff(R.string.fragment_settings_activation_show_logo_summary_off)
        showLogoPreference.isPersistent = false
        showLogoPreference.onPreferenceChangeListener = Preference.OnPreferenceChangeListener { _, newValue ->
            userSettings!!.load(activity)
            userSettings!!.showLogo = newValue as Boolean
            userSettings!!.save(activity)
            LauncherOverlayService.notifyConfigChanged(activity)
            true
        }
    }

    private fun getSideSummary(isOnRightSide: Boolean): Int {
        return if (isOnRightSide)
            R.string.fragment_settings_appearance_side_optionright_summary
        else
            R.string.fragment_settings_appearance_side_optionleft_summary
    }

    private fun getGravitySummary(gravity: LauncherGravity): Int {
        when (gravity) {
            LauncherGravity.Top -> {
                return R.string.fragment_settings_appearance_gravity_optiontop_summary
            }
            LauncherGravity.Center -> {
                return R.string.fragment_settings_appearance_gravity_optioncenter_summary
            }
            LauncherGravity.Bottom -> {
                return R.string.fragment_settings_appearance_gravity_optionbottom_summary
            }
        }
    }
}
